lib/ansible/modules/cloud/memset/memset_memstore_user.py

```diff
@@ -0,0 +1,334 @@
+  delegate_to: localhost
+'''
+
+RETURN = '''
+---
+memset_api:
+  description: Info from the Memset API
+  returned: when changed or state == present
+  type: complex
+  contains:
+    admin:
+      description: Whether the user is an admin.
+      returned: always
+      type: bool
+      sample: False
+    enabled:
+      description: Whether the user is enabled.
+      returned: always
+      type: bool
+      sample: False
+    username:
+      description: The username.
+      returned: always
+      type: string
+      sample: 'test'
+'''
+
+import re
+from ansible.module_utils.basic import AnsibleModule
+from ansible.module_utils.memset import memset_api_call
+
+
+def api_validation(args=None):
+    '''
+    Perform some validation which will be enforced by Memset's API (see:
+    https://www.memset.com/apidocs/methods_memstore.user.html)
+    '''
+    username_re = r'^[a-z0-9-\_]{1}[a-z0-9-\.\_]{1,49}$'
+    errors = dict()
+
+    if args['state'] == 'present' and not args['password']:
+        errors['password'] = "A password is required when state is present."
+    if not re.match(username_re, args['username'].lower()):
+        errors['username'] = "Username can only contain numbers, letters, dots, dashes and underscores, and can't start with a dot. Must be 50 chars or less."
+
+    if len(errors) > 0:
+        module.fail_json(failed=True, msg=errors)
+
+
+def create_user(args=None, user=None):
+    '''
+    Creates or updates a user.
+    '''
+    retvals, payload = dict(), dict()
+    retvals['changed'], retvals['failed'] = False, False
+    _set_password = False
+
+    if not user:
+        '''
+        User doesn't exist, create it. Creating a user also sets their password for the
+        first time, therefore respecting update_password when it is set to either `on_create`
+        or `always`.
+        '''
+        payload['name'], payload['username'], payload['password'], payload['enabled'] = \
+            args['memstore'], args['username'], args['password'], args['enabled']
+        if args['check_mode']:
+            retvals['changed'] = True
+            retvals['memset_api'] = payload
+            return(retvals)
+        # create the user
+        api_method = 'memstore.user.create'
+        retvals['failed'], msg, response = memset_api_call(api_key=args['api_key'], api_method=api_method, payload=payload)
+
+        if retvals['failed']:
+            retvals['msg'] = msg
+        else:
+            retvals['changed'] = True
+            retvals['memset_api'] = response.json()
+    else:
+        '''
+        User exists, update it. The only user attribute we can change is to enable or disable.
+        The password will be changed if update_password is set to `always`.
+        '''
+        if args['update_password'] == 'always':
+            _set_password = True
+
+        if user['enabled'] == args['enabled']:
+            if args['check_mode']:
+                return(retvals)
```

Comment:
IMHO this should be just as follows because check mode is not relevant if no action is taken.

~~~python
        if user['enabled'] == args['enabled']:
            return(retvals)
~~~

Reply:
There is a small difference: if `_set_password` is `True`, the module should set the password and not return here. Setting the password is done later in the code.

It's probably better to not return here at all, neither in check mode nor outside check mode, since otherwise `changed` is not set correctly. After all, `update_password` has check mode support built in.